include/wil/cppwinrt_helpers.h

```diff
@@ -487,11 +473,51 @@ namespace wil::details
         }
 
     private:
-        IterationStatus m_status{ IterationStatus::Initial };
-        winrt::array_view<TResult> m_values;
-        TResult* m_current{ nullptr };
+        IterationStatus m_status{ IterationStatus::Producing };
+        winrt::array_view<TResult> m_values{ &m_last_value, 1 };
+        TResult* m_current{ &m_last_value };
         TResult m_last_value{ empty<TResult>() };
     };
+
+    template<typename TResult, typename Func, typename... Args>
+    struct iterable_iterator_helper : winrt::implements<
+            iterable_iterator_helper<TResult, Func, Args...>,
+            winrt::Windows::Foundation::Collections::IIterable<TResult>
+            >
+    {
+        iterable_iterator_helper(Func&& func, Args&&... args) :
+            m_func{ std::forward<Func>(func) },
+            m_args{ std::forward<Args>(args)... }
+        {
+        }
+
+        auto First()
+        {
+            return std::apply(m_func, m_args);
+        }
+
+    private:
+        Func m_func;
+        std::tuple<Args...> m_args;
+    };
+
+    template<typename>
+    struct iterator_result;
+
+    template<typename TResult>
+    struct iterator_result<winrt::Windows::Foundation::Collections::IIterator<TResult>>
+    {
+        using type = TResult;
+    };
+}
+
+namespace wil
+{
+    template<typename Func, typename... Args, typename TResult = typename details::iterator_result<std::invoke_result_t<Func, Args...>>::type>
+    winrt::Windows::Foundation::Collections::IIterable<TResult> make_iterable_from_iterator(Func&& func, Args&&... args)
+    {
+        return winrt::make<details::iterable_iterator_helper<TResult, Func, Args...>>(std::forward<Func>(func), std::forward<Args>(args)...);

```

Comment:
I suggest adding a way to keep a strong or weak reference to a class here, as well as a pointer. Maybe something similar to C++/WinRT's delegate implementation.

Reply:
If `Func` accepts those arguments, it's already possible (e.g. calling a member function with a strong reference that is stored in the `IIterable<T>` helper with `wil::make_iterable_from_iterator(&Foo::Bar, get_strong())`. That doesn't work with weak references, but what would the semantics be if the weak pointer's object has been destroyed? Should `First` return an empty iterator?

Reply:
A member function pointer won't take a strong reference as argument directly via std::invoke/std::apply.

Good point reguarding weak references, lets just ignore that for now.

Reply:
[It will.](https://godbolt.org/z/xTsoG1v7s)

Reply:
Ah looks like std::invoke will try `(*arg).(*f)()`, I did not know that. Nothing to do here, then.